src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+                "frozen function.",
+                index, v->size, v->data, jitc_var_label(index));
+        }
+
+        jitc_raise("record(): Variable r%u[%u] -> %p, label=%s, data=%s, of "
+                   "size > 1 was created while recording.",
+                   index, v->size, v->data, jitc_var_label(index),
+                   jitc_var_str(index));
+
+        return 0;
+    }
+
+    /**
+     * Add information about a variable, deduplicating it and returning the slot
+     * in the `variables` field of the recording.
+     * Information is combined when the variable has already been added.
+     * This is used by the input variables of a kernel.
+     */
+    uint32_t add_variable(const void *ptr, RecordVariable rv) {
+
+        rv.ptr  = ptr;
+        auto it = this->ptr_to_slot.find(ptr);
+
+        if (it == this->ptr_to_slot.end()) {
+            uint32_t slot = this->m_recording.record_variables.size();
+
+            this->m_recording.record_variables.push_back(rv);
+
+            this->ptr_to_slot.insert({ ptr, slot });
+
+            return slot;
+        } else {
+            uint32_t slot = it.value();
+
+            this->m_recording.record_variables[slot] |= rv;
+
+            return slot;
+        }
+    }
+
+    /// Return the slot index given the data pointer of a variable.
+    /// This fails if the variable has not been added.
+    uint32_t get_variable(const void *ptr) {
+        auto it = this->ptr_to_slot.find(ptr);
+
+        if (it == this->ptr_to_slot.end())
+            jitc_fail("Failed to find the slot corresponding to the variable "
+                      "with data at %p",
+                      ptr);
+
+        return it.value();
+    }
+
+    /// Test if the ThreadState knows this \c ptr
+    bool has_variable(const void *ptr) {
+        auto it = this->ptr_to_slot.find(ptr);
+
+        return it != this->ptr_to_slot.end();
+    }
+
+    /**
+     * Adds a parameter access to the \ref dependencies vector.
+     * This also modifies the state of the \ref RecordVariable that was
+     * accessed.
+     */
+    void add_param(ParamInfo info) {
+        RecordVariable &rv = this->m_recording.record_variables[info.slot];
+        if (info.type == ParamType::Output) {
+
+            jitc_log(LogLevel::Debug, " <- param s%u", info.slot);
+
+            if (info.vtype != VarType::Void)
+                rv.type = info.vtype;
+
+            rv.state = RecordVarState::OpOutput;
+
+        } else if (info.type == ParamType::Input) {
+
+            jitc_log(LogLevel::Debug, " -> param s%u", info.slot);
+
+            if (info.test_uninit && rv.state == RecordVarState::Uninit)
+                jitc_raise("record(): Varaible at slot s%u was read from by "
+                           "operation o%u, but has not yet been initialized! "
+                           "This can happen if the variable was not part of "
+                           "the input but is used by an recorded operation.",
```

Comment:
Explain how the user could make it "part of the input", if possible.

Reply:
Not sure, does this make sense?
```c++
                jitc_raise("record(): Varaible at slot s%u was read from by "
                           "operation o%u, but has not yet been initialized! "
                           "This can occur if the variable was not part of "
                           "the input but is used by a recorded operation, for "
                           "example if it was not specified as a member in a "
                           "DRJIT_STRUCT but used in the frozen function.",
                           info.slot,
                           (uint32_t) this->m_recording.operations.size());
```

Reply:
Okay! Fixed a couple of typos below:

```cpp
                jitc_raise("record(): Variable at slot s%u was read by "
                           "operation o%u, but it had not yet been initialized! "
                           "This can occur if the variable was not part of "
                           "the input but is used by a recorded operation, for "
                           "example if it was not specified as a member in a "
                           "DRJIT_STRUCT but used in the frozen function.",
                           info.slot,
                           (uint32_t) this->m_recording.operations.size());
```